zerver/tests/test_auth_backends.py

```diff
@@ -1872,6 +1872,364 @@ def test_idp_initiated_signin_subdomain_implicit_invalid(self) -> None:
+    # failing because of assigning `url=self.LOGIN_URL` which gets
+    # changed because of `alternative_start_url` we send as a param
+    # and has no role in any test.
+    LOGIN_URL = ''
+
+    def prepare_login_url_and_headers(
+        self,
+        subdomain: Optional[str]=None,
+        mobile_flow_otp: Optional[str]=None,
+        desktop_flow_otp: Optional[str]=None,
+        is_signup: bool=False,
+        next: str='',
+        multiuse_object_key: str='',
+        alternative_start_url: Optional[str]=None,
+        *,
+        user_agent: Optional[str]=None,
+        skip_id_token: Optional[bool]=None,
+        account_data_dict: Optional[Dict[str, str]]=None,
+    ) -> Tuple[str, Dict[str, Any]]:
+        url, headers = super().prepare_login_url_and_headers(
+            subdomain, mobile_flow_otp, desktop_flow_otp, is_signup, next,
+            multiuse_object_key, alternative_start_url=self.AUTH_FINISH_URL,
+            user_agent=user_agent,
+        )
+
+        params = {'native_flow': 'true'}
+
+        if not skip_id_token and account_data_dict is not None:
+            params['id_token'] = self.generate_id_token(account_data_dict,
+                                                        settings.SOCIAL_AUTH_APPLE_BUNDLE_ID)
+
+        if is_signup:
+            params['is_signup'] = '1'
+
+        if subdomain:
+            params['subdomain'] = subdomain
+
+        url += "&%s" % (urllib.parse.urlencode(params),)
+        return url, headers
+
+    def social_auth_test(self, account_data_dict: Dict[str, str],
+                         *, subdomain: Optional[str]=None,
+                         mobile_flow_otp: Optional[str]=None,
+                         desktop_flow_otp: Optional[str]=None,
+                         is_signup: bool=False,
+                         next: str='',
+                         multiuse_object_key: str='',
+                         alternative_start_url: Optional[str]=None,
+                         user_agent: Optional[str]=None,
+                         **extra_data: Any) -> HttpResponse:
+        url, headers = self.prepare_login_url_and_headers(
+            subdomain, mobile_flow_otp, desktop_flow_otp, is_signup, next,
+            multiuse_object_key, alternative_start_url=self.AUTH_FINISH_URL,
+            user_agent=user_agent, skip_id_token=extra_data.get('skip_id_token'),
+            account_data_dict=account_data_dict,
+        )
+
+        with responses.RequestsMock(assert_all_requests_are_fired=False) as requests_mock:
+            # The below mock is to get public key we generated
+            # instead of getting it from
+            # https://appleid.apple.com/auth/keys.
+            requests_mock.add(
+                requests_mock.GET,
+                self.BACKEND_CLASS.JWK_URL,
+                status=200,
+                json=json.loads(settings.APPLE_JWK),
+            )
+            result = self.client_get(url, **headers)
```

Comment:
Why is this here, rather than in `prepare_login_url_and_headers`?  I'm also a little puzzled why this seems to not actually call the upstream `social_auth_test` method -- if intentional, it certainly deserves a big comment, because it means we're not running most of our test code.

Reply:
Because we want to test the native flow code, which is just sending `GET` request with `native_flow=true` and the other parts of upstream `social_auth_test` are already tested, we start the flow directly with `result = self.client_get(url, **headers)`. We have it there instead of `prepare_login_url_headers` as we need to mock the request requiring internet that this auth flow makes to get apple public JWK from `https://appleid.apple.com/auth/keys.`.

Reply:
Added a comment explaining a bit about native flow and why we start with `self.client_get(url, **headers)`.
We can have the remaining code also but we don't really focus on that in testing native flow and it's tested already in `AppleIdAuthBackendTest`.